clippy_lints/src/cargo_common_metadata.rs

```diff
@@ -80,32 +93,34 @@ impl LateLintPass<'_> for CargoCommonMetadata {
         let metadata = unwrap_cargo_metadata!(cx, CARGO_COMMON_METADATA, false);
 
         for package in metadata.packages {
-            if is_empty_vec(&package.authors) {
-                missing_warning(cx, &package, "package.authors");
-            }
-
-            if is_empty_str(&package.description) {
-                missing_warning(cx, &package, "package.description");
-            }
-
-            if is_empty_str(&package.license) && is_empty_path(&package.license_file) {
-                missing_warning(cx, &package, "either package.license or package.license_file");
-            }
-
-            if is_empty_str(&package.repository) {
-                missing_warning(cx, &package, "package.repository");
-            }
-
-            if is_empty_path(&package.readme) {
-                missing_warning(cx, &package, "package.readme");
-            }
-
-            if is_empty_vec(&package.keywords) {
-                missing_warning(cx, &package, "package.keywords");
-            }
-
-            if is_empty_vec(&package.categories) {
-                missing_warning(cx, &package, "package.categories");
+            if package.publish.as_ref().filter(|publish| publish.is_empty()).is_none() || self.ignore_publish {
```

Comment:
Shouldn't you check if `publish = false`? A check for `publish.is_none()` would bean, that it should only lint, if no publish attribute exists. So `publish = true` would also not get linted.

Reply:
This check isn't really easy to read. I will add some documentation.
[The documentation says](https://docs.rs/cargo_metadata/0.12.3/cargo_metadata/struct.Package.html#structfield.publish):
> Publishing is unrestricted if None, and forbidden if the Vec is empty.

- Cases we want to lint for, because they allow publishing:
  - nothing -> `None`
  - `publish = true` -> `None`
  - `publish = ["somewhere"]` -> `Some(vec!["somewhere"])`
- Cases we do **NOT** want to lint for, because they do **NOT** allow publishing:
  - `publish = false` -> `Some(vec![])`
  - `publish = []` -> `Some(vec![])`

So basically we want to lint if it's none or the `Vec` isn't empty.

Reply:
https://github.com/rust-lang/rust-clippy/blob/fd8b5fa1aa614dcfd21471ab5a9a56c7873ce14a/clippy_lints/src/cargo_common_metadata.rs#L96-L98

I hope this is alright?